repose-aggregator/components/services/opentracing-service/opentracing-service-impl/src/main/scala/org/openrepose/core/services/opentracing/TracerExtractor.scala

```diff
@@ -0,0 +1,105 @@
+ */
+package org.openrepose.core.services.opentracing
+
+import java.util
+import java.util.AbstractMap.SimpleImmutableEntry
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import io.opentracing.propagation.TextMap
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+
+import scala.collection.mutable.ListBuffer
+
+class TracerExtractor() extends TextMap with LazyLogging {
+  var headers: Map[String, List[String]] = _
+
+  def this(httpServletRequestWrapper: HttpServletRequestWrapper) = {
+    this()
+
+    headers = servletHeadersToMultiMap(httpServletRequestWrapper)
+
+    def servletHeadersToMultiMap(httpServletRequest: HttpServletRequestWrapper): Map[String, List[String]] = {
+      val headersResult = scala.collection.mutable.HashMap.empty[String, List[String]]
+      if (httpServletRequest != null) {
+        val headerNamesIterator = httpServletRequest.getHeaderNames
+        while ( {
+          headerNamesIterator.hasMoreElements
+        }) {
+          val headerName: String = headerNamesIterator.nextElement
+          val valuesIterator = httpServletRequest.getHeaders(headerName)
+          var valuesList = ListBuffer[String]()
+          while ( {
+            valuesIterator.hasMoreElements
+          }) {
+            valuesList += valuesIterator.nextElement
+          }
+          headersResult += (headerName -> valuesList.toList)
+        }
+      }
+
+      headersResult.toMap
+
+    }
+  }
+
+  override def put(key: String, value: String): Unit = ???
+
+  override def iterator() = new MultivaluedMapFlatIterator[String, String](this.headers.toSet)
+
+  /**
+    * Iterate through a map with lists
+    * @tparam K String key
+    * @tparam V List[String] value
+    */
+  class MultivaluedMapFlatIterator[K <: AnyRef, V >: Null <: AnyRef]() extends util.Iterator[util.Map.Entry[K, V]] {
+    var mapIterator: Iterator[(K, List[V])] = _
+    var mapEntry: (K, List[V]) = _
+    var listIterator: Iterator[V] = _
+
+    def this(multiValuesEntrySet: Set[(K, List[V])]) = {
+      this()
+
+      mapIterator = multiValuesEntrySet.iterator
+      println(s"ctor $mapIterator")
+    }
+
+    override def hasNext: Boolean = {
+      println(s"hasNext $listIterator")
+      if (listIterator != null && listIterator.hasNext) return true
+      println(s"hasNext ${mapIterator.hasNext}")
+      mapIterator.hasNext
+    }
+
+    override def next: util.Map.Entry[K, V] = {
+      println(s"next $mapEntry")
+      if (mapEntry == null || (!listIterator.hasNext && mapIterator.hasNext)) {
+        mapEntry = mapIterator.next
+        println(s"next 2 $mapEntry")
+        listIterator = mapEntry._2.iterator
+        println(s"listIterator $listIterator")
+      }
+      println(s"listIterator 2 ${listIterator.hasNext}")
+      if (listIterator.hasNext) new SimpleImmutableEntry[K, V](mapEntry._1, listIterator.next)
+      else new SimpleImmutableEntry[K, V](mapEntry._1, null)
+    }
+
+  }
+}
```

Comment:
Here's my rewrite of your class!
```
class TracerExtractor(httpServletRequestWrapper: HttpServletRequestWrapper) extends TextMap with LazyLogging {

  val headers: Map[String, List[String]] = Option(httpServletRequestWrapper).map(request =>
    request.getHeaderNamesScala
      .map(headerName => headerName -> request.getHeadersScala(headerName))
      .toMap
  ).getOrElse(Map.empty)

  override def put(key: String, value: String): Unit = ???

  override def iterator(): util.Iterator[util.Map.Entry[String, String]] = {
    headers.toIterator
      .flatMap(tupled((key, values) => values.map(key -> _)))
      .map(tupled((key, value) => new util.AbstractMap.SimpleEntry(key, value)))
      .asJava
      .asInstanceOf[util.Iterator[util.Map.Entry[String, String]]]
  }
}
```

Reply:
As a side note, that `.asInstanceOf` line was annoying. Despite `SimpleEntry` being an implementation of the `Map.Entry` interface, the typing is not correct because:
> Note: java.util.AbstractMap.SimpleEntry[String,String] <: java.util.Map.Entry[String,String], but Java-defined trait Iterator is invariant in type E.
> You may wish to investigate a wildcard type such as `_ <: java.util.Map.Entry[String,String]`. (SLS 3.2.10)